grpcclient/grpcclient.go

```diff
@@ -0,0 +1,104 @@
+}
+
+// RegisterFlagsWithPrefix registers flags with prefix.
+func (cfg *Config) RegisterFlagsWithPrefix(prefix string, f *flag.FlagSet) {
+	f.IntVar(&cfg.MaxRecvMsgSize, prefix+".grpc-max-recv-msg-size", 100<<20, "gRPC client max receive message size (bytes).")
+	f.IntVar(&cfg.MaxSendMsgSize, prefix+".grpc-max-send-msg-size", 16<<20, "gRPC client max send message size (bytes).")
+	f.StringVar(&cfg.GRPCCompression, prefix+".grpc-compression", "", "Use compression when sending messages. Supported values are: 'gzip', 'snappy' and '' (disable compression)")
+	f.Float64Var(&cfg.RateLimit, prefix+".grpc-client-rate-limit", 0., "Rate limit for gRPC client; 0 means disabled.")
+	f.IntVar(&cfg.RateLimitBurst, prefix+".grpc-client-rate-limit-burst", 0, "Rate limit burst for gRPC client.")
+	f.BoolVar(&cfg.BackoffOnRatelimits, prefix+".backoff-on-ratelimits", false, "Enable backoff and retry when we hit ratelimits.")
+	f.BoolVar(&cfg.TLSEnabled, prefix+".tls-enabled", cfg.TLSEnabled, "Enable TLS in the GRPC client. This flag needs to be enabled when any other TLS flag is set. If set to false, insecure connection to gRPC server will be used.")
+
+	cfg.BackoffConfig.RegisterFlagsWithPrefix(prefix, f)
+
+	cfg.TLS.RegisterFlagsWithPrefix(prefix, f)
+}
+
+func (cfg *Config) Validate(log log.Logger) error {
+	switch cfg.GRPCCompression {
+	case gzip.Name, snappy.Name, "":
+		// valid
+	default:
+		return errors.Errorf("unsupported compression type: %s", cfg.GRPCCompression)
+	}
+	return nil
+}
+
+// CallOptions returns the config in terms of CallOptions.
+func (cfg *Config) CallOptions() []grpc.CallOption {
+	var opts []grpc.CallOption
+	opts = append(opts, grpc.MaxCallRecvMsgSize(cfg.MaxRecvMsgSize))
+	opts = append(opts, grpc.MaxCallSendMsgSize(cfg.MaxSendMsgSize))
+	if cfg.GRPCCompression != "" {
+		opts = append(opts, grpc.UseCompressor(cfg.GRPCCompression))
+	}
+	return opts
+}
+
+// DialOption returns the config as a grpc.DialOptions.
+func (cfg *Config) DialOption(unaryClientInterceptors []grpc.UnaryClientInterceptor, streamClientInterceptors []grpc.StreamClientInterceptor) ([]grpc.DialOption, error) {
+	var opts []grpc.DialOption
+	tlsOpts, err := cfg.TLS.GetGRPCDialOptions(cfg.TLSEnabled)
+	if err != nil {
+		return nil, err
+	}
+	opts = append(opts, tlsOpts...)
+
+	if cfg.BackoffOnRatelimits {
+		unaryClientInterceptors = append([]grpc.UnaryClientInterceptor{NewBackoffRetry(cfg.BackoffConfig)}, unaryClientInterceptors...)
+	}
+
+	if cfg.RateLimit > 0 {
+		unaryClientInterceptors = append([]grpc.UnaryClientInterceptor{NewRateLimiter(cfg)}, unaryClientInterceptors...)
+	}
+
+	return append(
+		opts,
+		grpc.WithDefaultCallOptions(cfg.CallOptions()...),
+		grpc.WithUnaryInterceptor(grpc_middleware.ChainUnaryClient(unaryClientInterceptors...)),
+		grpc.WithStreamInterceptor(grpc_middleware.ChainStreamClient(streamClientInterceptors...)),
+		grpc.WithKeepaliveParams(keepalive.ClientParameters{
+			Time:                time.Second * 20,
```

Comment:
These very low ping times proved problematic, because they require correct server-side settings too. I think we should at least document that. See https://github.com/cortexproject/cortex/issues/3606#issuecomment-745179803 for details.

Reply:
@pstibrany I made these into parameters `PingTime` and `PingTimeout`. I didn't give them any default values, thinking that maybe it's best to use the `gRPC` defaults and let applications choose their own. Does that make sense to you, or should we use respectively 20 and 10 seconds for defaults?

Reply:
Thanks @pstibrany, let's see what Marco says.

Reply:
My take on this topic:
1. This is quite a deep thing and I wouldn't let it configurable because will probably just pollute the config more than we do right now
2. Doing logic changes while porting features from Cortex to dskit makes review way more difficult. Even if we'll decide to do changes, these should be done in a separate PR in my opinion.
3. Even if we add it to the config, I would use the same naming used by grpc (eg. keepalive instead of ping, ...)